providers/dns/azion/azion.go

```diff
@@ -0,0 +1,279 @@
+	TTL                int
+	HTTPClient         *http.Client
+}
+
+// NewDefaultConfig returns a default configuration for the DNSProvider.
+func NewDefaultConfig() *Config {
+	return &Config{
+		TTL:                env.GetOrDefaultInt(EnvTTL, dns01.DefaultTTL),
+		PropagationTimeout: env.GetOrDefaultSecond(EnvPropagationTimeout, dns01.DefaultPropagationTimeout),
+		PollingInterval:    env.GetOrDefaultSecond(EnvPollingInterval, dns01.DefaultPollingInterval),
+		HTTPClient: &http.Client{
+			Timeout: env.GetOrDefaultSecond(EnvHTTPTimeout, 30*time.Second),
+		},
+	}
+}
+
+// DNSProvider implements the challenge.Provider interface.
+type DNSProvider struct {
+	config *Config
+	client *idns.APIClient
+
+	recordIDs   map[string]int32
+	recordIDsMu sync.Mutex
+}
+
+// NewDNSProvider returns a DNSProvider instance configured for Azion.
+// Credentials must be passed in the environment variable: AZION_PERSONAL_TOKEN.
+func NewDNSProvider() (*DNSProvider, error) {
+	values, err := env.Get(EnvPersonalToken)
+	if err != nil {
+		return nil, fmt.Errorf("azion: %w", err)
+	}
+
+	config := NewDefaultConfig()
+	config.PersonalToken = values[EnvPersonalToken]
+
+	return NewDNSProviderConfig(config)
+}
+
+// NewDNSProviderConfig return a DNSProvider instance configured for Azion.
+func NewDNSProviderConfig(config *Config) (*DNSProvider, error) {
+	if config == nil {
+		return nil, errors.New("azion: the configuration of the DNS provider is nil")
+	}
+
+	if config.PersonalToken == "" {
+		return nil, errors.New("azion: missing credentials")
+	}
+
+	clientConfig := idns.NewConfiguration()
+	clientConfig.AddDefaultHeader("Accept", "application/json; version=3")
+	clientConfig.UserAgent = "lego-dns/azion"
+
+	if config.HTTPClient != nil {
+		clientConfig.HTTPClient = config.HTTPClient
+	}
+
+	client := idns.NewAPIClient(clientConfig)
+
+	return &DNSProvider{
+		config:    config,
+		client:    client,
+		recordIDs: make(map[string]int32),
+	}, nil
+}
+
+// Timeout returns the timeout and interval to use when checking for DNS propagation.
+func (d *DNSProvider) Timeout() (timeout, interval time.Duration) {
+	return d.config.PropagationTimeout, d.config.PollingInterval
+}
+
+// Present creates a TXT record using the specified parameters.
+func (d *DNSProvider) Present(domain, token, keyAuth string) error {
+	info := dns01.GetChallengeInfo(domain, keyAuth)
+
+	zone, err := d.findZone(info.EffectiveFQDN)
+	if err != nil {
+		return fmt.Errorf("azion: could not find zone for domain %q: %w", domain, err)
+	}
+
+	subDomain, err := dns01.ExtractSubDomain(info.EffectiveFQDN, zone.GetName())
+	if err != nil {
+		return fmt.Errorf("azion: %w", err)
+	}
+
+	ctx := context.WithValue(context.Background(), idns.ContextAPIKeys, map[string]idns.APIKey{
+		"tokenAuth": {
+			Key:    d.config.PersonalToken,
+			Prefix: "Token",
+		},
+	})
+
+	// For root domains, the subdomain will be empty, so we use _acme-challenge directly
+	if subDomain == "" {
+		subDomain = "_acme-challenge"
+	}
```

Comment:
This will not work as expected with CNAME.

`info.EffectiveFQDN` is the domain after the CNAME evaluation doesn't have to use `_acme-challenge` prefix.

https://letsencrypt.org/2019/10/09/onboarding-your-customers-with-lets-encrypt-and-acme/#the-advantages-of-a-cname

Reply:
Fixed 👊🏻

Reply:
`info.EffectiveFQDN` must be used.

The problem is not the usage of `info.EffectiveFQDN`.

Reply:
What are the possible values for `entry`?
What is the `entry` when the domain is a zone (maybe `@`)?

Reply:
based on the API documentation it's `@`

Reply:
Yes, it's `@`. Thanks for the updates.